### DISCUSSION TITLE
Two CF-Cards in RomWBW

### DISCUSSION THREAD

**Discussion by hubertushirsch** [wwarthen/RomWBW#454] [Q&A] [answered]
Hello Wayne,

Do you have any experience with using two CF cards in RomWBW?

I have an SC720 Z80 SBC with a built-in IDE-CF interface and an SC126 Z180 SBC, extended with an SC729 module.
Different cards work in both systems without problems.

Now I tried to use 2 CF cards in the same system.
I also plugged the SC729 into the SC720 and jumpered one adapter to master and the other to slave.

Is that the right way?

I get very different results when I swap cards or swap master and slave.
Only a few combinations work, usually only IDE0 is recognized, on IDE1 No Media in RomWBW

I also found that there are dual CF-IDE adapters on the market. e.g. here:
https://usb.brando.com/dual-cf-to-ide-adapter_p00650c0042d015.html
Do they work with 2 CF cards in RomWBW?

Do you have any experience with using two CF cards?

I already asked Steve Cousins, he wrote to me:
*****************
I tried similar ideas and was unsuccessful. I believe this is normal behaviour. If I remember correctly I decided that two cards (one master, and one slave) would not work with the addressing scheme used by my module and other similar modules.

After a little digging I think RomWBW can be configured to support two Compact Flash interfaces with different I/O addresses and presumably both devices set to master mode. I have not tried this myself but I've seen mention on forums of people successfully connecting two Compact Flash modules. It is necessary to modify one of them to the required I/O address (0x20 if I'm remembering correctly) and build a custom RomWBW configuration. My memory may be defective here, but I think this is the case.

Sorry, I don't have a more definitive answer for you.

Regards,
Steve
*****************
Regards
Hubert

**Comment by wwarthen** [comment]
Hi Hubert,

RomWBW definitely does support multiple IDE interfaces.  Further, it supports both a master and a space device on each interface.

I am currently traveling, so I'm not able to look up the configuration specifics of your module.  In general, RomWBW must be properly configured for each IDE device in your system.  CF Cards are IDE devices.  RomWBW will automatically look for both the master and slave with no additional configuration.

Your success will depend somewhat on the specific CF Cards.  Some CF Cards do not play well with others.  If there is only one CF Card on an interface, it should be configured as master.  Slave without master may work, but depends on the card and is not recommended by the IDE spec.

I will review the specifics of your modules in the next day or so and try to provide more help.

Thanks,

Wayne

**Reply by wwarthen** [reply]
I forgot to address one of your specific questions.

Yes, a dual CF Adapter will normally work and is fine under RomWBW.  That is the easiest way to add another CF Card.  As long as the dual adapter fits, it should just work (noting my prior caveats about some CF Cards not playing well together).  No additional configuration of RomWBW is required for it to detect the second (slave) card on an existing interface.

Thanks,

Wayne

**Reply by hubertushirsch** [reply]
Hi Wayne,

thanks for the answer. 
I have just ordered a dual CF adapter and will test it next week when it arrives.
I will report back on the results.

Hubert

**Comment by wwarthen** [comment]
Sounds good Hubert.

I have reviewed the relevant schematics and it looks like this should work.

Thanks, Wayne

**Comment by hubertushirsch** [comment, marked as answer]
The DUAL CF adapters arrived today.
It works as I wanted it to.
All CF cards that previously worked solo also work in any combination with each other.
The dual CF adapter saves me the space for a second IDE interface.

Great.
Hubert
![sc720_with_two_cf_cards](https://github.com/user-attachments/assets/ff85f5b7-7529-45a1-a6fb-1bdf59a1b23e)
![sc126_with_two_cf_cards](https://github.com/user-attachments/assets/a7141801-bf36-4793-9355-edeb1755bedf)

```
RomWBW HBIOS v3.5.0-dev.91, 2024-10-13

Small Computer SC126 [SCZ180_sc126_std] Z8S180-N @ 18.432MHz IO=0xC0
0 MEM W/S, 2 I/O W/S, INT MODE 2, Z180 MMU
512KB ROM, 512KB RAM, HEAP=0x1F82
ROM VERIFY: 00 00 00 00 PASS

LCD: IO=0xAA NOT PRESENT
AY: MODE=RCZ180 IO=0x68 NOT PRESENT
ASCI0: IO=0xC0 ASCI W/BRG MODE=115200,8,N,1
ASCI1: IO=0xC1 ASCI W/BRG MODE=115200,8,N,1
DSRTC: MODE=STD IO=0x0C Mon 2024-11-18 17:30:16 CHARGE=OFF
MD: UNITS=2 ROMDISK=384KB RAMDISK=256KB
FD: MODE=RCWDC IO=0x50 NOT PRESENT
IDE: IO=0x10 MODE=RC
IDE0: ATA 8-BIT LBA BLOCKS=0x0003D400 SIZE=122MB
IDE1: ATA 8-BIT LBA BLOCKS=0x0007A2B0 SIZE=244MB
PPIDE: IO=0x20 PPI NOT PRESENT
SD: MODE=SC OPR=0x0C CNTR=0xCA TRDR=0xCB DEVICES=1
SD0: SDHC NAME=SU04G BLOCKS=0x00762C00 SIZE=3781MB
CH0: IO=0x3E NOT PRESENT
CH1: IO=0x3C NOT PRESENT
FP: IO=0x00 NOT PRESENT

Unit        Device      Type              Capacity/Mode
----------  ----------  ----------------  --------------------
Char 0      ASCI0:      RS-232            115200,8,N,1
Char 1      ASCI1:      RS-232            115200,8,N,1
Disk 0      MD0:        RAM Disk          256KB,LBA
Disk 1      MD1:        ROM Disk          384KB,LBA
Disk 2      IDE0:       Hard Disk         122MB,LBA
Disk 3      IDE1:       CompactFlash      244MB,LBA
Disk 4      SD0:        SD Card           3781MB,LBA
```

**Reply by wwarthen** [reply]
Excellent.  This is exactly as it should be.

Thanks, Wayne
